certbot/compat/filesystem.py

```diff
@@ -312,6 +314,109 @@ def is_executable(path):
+    }))
+
+
+def compute_private_key_mode(old_key, base_mode):
+    # type: (str, int) -> int
+    """
+    Calculate the POSIX mode to apply to a private key given the previous private key
+    :param str old_key: path to the previous private key
+    :param int base_mode: the minimum modes to apply to a private key
+    :return: the POSIX mode to apply
+    :rtype: int
+    """
+    if POSIX_MODE:
+        # On Linux, we keep read/write/execute permissions
+        # for group and read permissions for everybody.
+        old_mode = (stat.S_IMODE(os.stat(old_key).st_mode) &
+                    (stat.S_IRGRP | stat.S_IWGRP | stat.S_IXGRP | stat.S_IROTH))
+        return base_mode | old_mode
+
+    # On Windows, the mode returned by os.stat is not reliable,
+    # so we do not keep any permission from the previous private key.
+    return base_mode
+
+
+def compare_ownership(path1, path2):
+    # type: (str, str) -> bool
+    """
+    Compare the ownership of two files given their respective path
+    :param str path1: path to the first file
+    :param str path2: path to the second file
+    :return: True if both files have the same ownership, False otherwise
+    :rtype: bool
+
+    """
+    if POSIX_MODE:
+        stats1 = os.stat(path1)
+        stats2 = os.stat(path2)
+        return (stats1.st_uid, stats1.st_gid) == (stats2.st_uid, stats2.st_gid)
+
+    security1 = win32security.GetFileSecurity(path1, win32security.OWNER_SECURITY_INFORMATION)
+    user1 = security1.GetSecurityDescriptorOwner()
+
+    security2 = win32security.GetFileSecurity(path2, win32security.OWNER_SECURITY_INFORMATION)
+    user2 = security2.GetSecurityDescriptorOwner()
+
+    return user1 == user2
+
+
+def has_min_permissions(path, min_mode):
+    # type: (str, int) -> bool
+    """
+    Check if a file given its path has at least the permissions defined by the given minimal mode.
+    :param str path: path to the file to check
+    :param int min_mode: the minimal permissions expected
+    :return: True if the file matches the minimal permissions expectations, False otherwise
+    :rtype: bool
+    """
+    if POSIX_MODE:
+        st_mode = os.stat(path).st_mode
+        return st_mode == st_mode | min_mode
+
+    path = realpath(path)
+
+    # Get owner sid of the file
+    security = win32security.GetFileSecurity(
+        path, win32security.OWNER_SECURITY_INFORMATION | win32security.DACL_SECURITY_INFORMATION)
+    user = security.GetSecurityDescriptorOwner()
+    dacl = security.GetSecurityDescriptorDacl()
+    min_dacl = _generate_dacl(user, min_mode)
+
+    for index in range(min_dacl.GetAceCount()):
+        min_ace = min_dacl.GetAce(index)
+        target_aces = [dacl.GetAce(index2) for index2 in range(dacl.GetAceCount())
+                      if dacl.GetAce(index2)[2] == min_ace[2]]
+        if not target_aces:
+            return False
+        if not target_aces[0][1] == target_aces[0][1] | min_ace[1]:
```

Comment:
With the files created by our `certbot.compat.filesystem` module, no. Because inserting a new ACE replaces the existing ACE for this SID, from the tests I did.

However, a second ACE can exist in files outside certbot, because the file gets permission inheritance from the parent directory. In this case you can have the ACE inherited, and the ACE defined explicitly.

So for the sake of completeness, I change the code to handle that case.

Reply:
Is the new logic what Windows does or is it more complex than that? Based on https://docs.microsoft.com/en-us/windows/win32/secauthz/how-dacls-control-access-to-an-object, I'd think the full logic would combine all ACEs for the SID, taking both access-allowed and access-denied into account based on the order of entries in the ACL.

Reply:
Yes it is! My implementation is too basic. I was writing the logic with the meta set of permission bits, not the bits themselves. You are right about the logic. I will add on top of this that in theory ACEs for the groups of the user must be taken into account.

So it is quite complex. The simplest thing to do I think is to let Windows calculate the effective permissions for the user, using the dedicated API, then make our assertions regarding the meta sets on the resulted mask.

Reply:
> The simplest thing to do I think is to let Windows calculate the effective permissions for the user, using the dedicated API, then make our assertions regarding the meta sets on the resulted mask.

Good idea! That definitely seems like the best approach.

Actually handling this correctly may be overkill because this function may only be called on files whose permissions are set by Certbot, but doing it right doesn't seem that hard and could avoid problems in the future.

Reply:
This is my thought also. But relying on the effective permission resolution is harmless for files that follows strictly our security model, and let `has_min_permissions` to return relevant results for files that do not.

Reply:
So, implemented.